src/Framework/Http/ConnectServer/Client/ConnectClient.php

```diff
@@ -0,0 +1,131 @@
+<?php
+
+namespace Give\Framework\Http\ConnectServer\Client;
+
+use Give\Framework\Http\ConnectServer\Client\Exceptions\RequestException;
+use WP_Error;
+
+/**
+ * Class ConnectClient
+ *
+ * @since   2.8.0
+ */
+class ConnectClient
+{
+    /**
+     * Api Url
+     *
+     * @since 2.8.0
+     *
+     * @var string
+     */
+    public $apiUrl = 'https://connect.givewp.com';
+
+    /**
+     * @unreleased
+     */
+    public function __construct($giveConnectUrl = '')
+    {
+        if ( ! empty($giveConnectUrl)) {
+            $this->apiUrl = $giveConnectUrl;
+        } elseif (defined('GIVE_CONNECT_URL') && GIVE_CONNECT_URL) {
+            $this->apiUrl = GIVE_CONNECT_URL;
```

Comment:
We should probably setup a singleton in the container that injects the value from the constant. This would decouple the class from the existence of the constant and improve test-ability.

Reply:
@kjohnson Thanks for the review, but I didn't understand why this is necessary, by default this const should not exist and it should use the [**$apiUrl default value**](https://github.com/impress-org/givewp/blob/feature/gateway-server-client-6680/src/Framework/Http/ConnectServer/Client/ConnectClient.php#L22). Basically, this const should be used just for local development, for test-ability with different URLs we can pass it as a param, like this: `$client = new ConnectClient('my-custom-url');`

Maybe we should rename it to `GIVE_DEV_CONNECT_URL` to let it more clear?

Reply:
Generally, my point is that the configuration should be injected into the class - so that the class is not concerned with determining which URL to use.

The use of a constant is configuring application state, which is specific to the WordPress environment. This can make it harder to test the class in isolation. Instead, if the URL is passed into the class then it can be tested more easily.

This isn't strictly necessary and is a recommendation. I don't want to hold up the PR over what is simply a refactor.

Reply:
I realize now how my original suggestion would be unclear. My apologies for rushing through that.

Your refactor moving the `ConnectClient` to a singleton looks good. The only change that I would make would be to make the `$giveConnectUrl` constructor argument required and pass the default in from the service container.

This creates a single source of truth for where the URL is being set, which would make debugging much easier.



```php
        give()->singleton(ConnectClient::class, static function () {
            $giveConnectUrl = (defined('GIVE_CONNECT_URL') && GIVE_CONNECT_URL)
                ? GIVE_CONNECT_URL
                : 'https://connect.givewp.com';

            return new ConnectClient($giveConnectUrl);
        });
```

```php
    public $apiUrl;

    public function __construct($giveConnectUrl)
    {
        $this->apiUrl = $giveConnectUrl;
    }
```

Reply:
@kjohnson No problem! Thanks for the extra improvement suggestions. Resolved in fe0ec8b1548249b520d1a5d89e7eca9bb959138f